### PR TITLE
Do not load CPtr

### PR DESCRIPTION
**Pull request by certik**

### CODE REVIEW

**Comment by anutosh491** [comment]
This seems to be the correct fix for the concerned program 
```
from lpython import S
from sympy import Symbol

def mmrv(x: S) -> list[S]:
    l1: list[S] = [x]
    return l1

def test_mrv1():
    x: S = Symbol("x")
    ans: list[S] = mmrv(x)

test_mrv1()
```
But fails for `symbolics_11.py` (actually just the first 2 lines)

**Comment by anutosh491** [comment]
Consider the following 
```
from sympy import pi
from lpython import S

def test_extraction_of_elements():
    l1: list[S] = [pi]

test_extraction_of_elements()
```
The above change would disturb this simple block that works on master. I think there's a type mismatch happens while storing elements into the `list`
```
define void @__module___main___test_extraction_of_elements() {
.entry:
  %_stack0 = alloca i64, align 8
  %l1 = alloca %list, align 8
  %0 = call i8* (i32, ...) @_lfortran_malloc(i32 8)
  %1 = bitcast i8* %0 to void**
  %2 = getelementptr %list, %list* %l1, i32 0, i32 2
  store void** %1, void*** %2, align 8
  %3 = getelementptr %list, %list* %l1, i32 0, i32 0
  store i32 0, i32* %3, align 4
  %4 = getelementptr %list, %list* %l1, i32 0, i32 1
  store i32 1, i32* %4, align 4
  %stack0 = alloca void*, align 8
  store i64 0, i64* %_stack0, align 4
  store void* null, void** %stack0, align 8
  %5 = bitcast i64* %_stack0 to void*
  store void* %5, void** %stack0, align 8
  %6 = load void*, void** %stack0, align 8
  call void @basic_new_stack(void* %6)
  %7 = load void*, void** %stack0, align 8
  call void @basic_const_pi(void* %7)
  %const_list = alloca %list, align 8
  %8 = call i8* (i32, ...) @_lfortran_malloc(i32 8)
  %9 = bitcast i8* %8 to void**
  %10 = getelementptr %list, %list* %const_list, i32 0, i32 2
  store void** %9, void*** %10, align 8
  %11 = getelementptr %list, %list* %const_list, i32 0, i32 0
  store i32 1, i32* %11, align 4
  %12 = getelementptr %list, %list* %const_list, i32 0, i32 1
  store i32 1, i32* %12, align 4
  %13 = getelementptr %list, %list* %const_list, i32 0, i32 2
  %14 = load void**, void*** %13, align 8
  %15 = getelementptr inbounds void*, void** %14, i32 0
  store void** %stack0, void** %15, align 8
  %16 = getelementptr %list, %list* %const_list, i32 0, i32 0
  %17 = load i32, i32* %16, align 4
  %18 = getelementptr %list, %list* %const_list, i32 0, i32 1
  %19 = load i32, i32* %18, align 4
  %20 = getelementptr %list, %list* %l1, i32 0, i32 0
  %21 = getelementptr %list, %list* %l1, i32 0, i32 1
  store i32 %17, i32* %20, align 4
  store i32 %19, i32* %21, align 4
  %22 = getelementptr %list, %list* %const_list, i32 0, i32 2
  %23 = load void**, void*** %22, align 8
  %24 = mul i32 8, %19
  %25 = call i8* (i32, ...) @_lfortran_malloc(i32 %24)
  %26 = bitcast i8* %25 to void**
  %27 = bitcast void** %26 to i8*
  %28 = bitcast void** %23 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i32(i8* %27, i8* %28, i32 %24, i1 false)
  %29 = getelementptr %list, %list* %l1, i32 0, i32 2
  store void** %26, void*** %29, align 8
  %30 = load void*, void** %stack0, align 8
  call void @basic_free_stack(void* %30)
  br label %return

return:                                           ; preds = %.entry
  ret void
}
```
The error 
```
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store void** %stack0, void** %15, align 8
 void*
```

**Comment by anutosh491** [comment]
I tried some workarounds to support both but unfortunately both the cases contradict each other.

**Comment by certik** [comment]
The failure file has the following relevant ASR (after all passes):
```cl
                                           l1:
                                                (Variable
                                                    3
                                                    l1
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (List
                                                        (CPtr)
                                                    )
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                ),
...
                                   (=
                                        (Var 3 l1)
                                        (ListConstant
                                            [(Var 3 stack0)]
                                            (List
                                                (CPtr)
                                            )
                                        )
                                        ()
                                    )
```
We are assigning a list of CPtr. The ASR looks correct, but the generated code is incorrect.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -1245,8 +1245,12 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         llvm::Value* const_list = builder->CreateAlloca(const_list_type, nullptr, "const_list");
         list_api->list_init(type_code, const_list, *module, x.n_args, x.n_args);
         int64_t ptr_loads_copy = ptr_loads;
-        ptr_loads = 1;
         for( size_t i = 0; i < x.n_args; i++ ) {
+            if (ASR::is_a<ASR::CPtr_t>(*ASRUtils::expr_type(x.m_args[i]))) {
+                ptr_loads = 0;
+            } else {
+                ptr_loads = 1;
+            }
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:1253]: I think the issue here is that if `x.m_args[i]` is an argument, then we want `ptr_loads=0`, but if it is `pi` (as in the other example), where it is a local (or global) variable, we want `ptr_loads = 1`.

**Comment by certik** [comment]
Good:
```
  %9 = bitcast i8* %8 to void**
  %10 = getelementptr %list, %list* %const_list, i32 0, i32 2
  store void** %9, void*** %10, align 8
```
bad:
```
  %14 = load void**, void*** %13, align 8
  %15 = getelementptr inbounds void*, void** %14, i32 0
  store void** %stack0, void** %15, align 8
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -1245,8 +1245,32 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                        ASR::Variable_t *v = down_cast<ASR::Variable_t>(var_sym);
+                        if (v->m_intent == intent_local ||
+                            v->m_intent == intent_return_var ||
+                            !v->m_intent) {
+                            is_argument = false;
+                        } else {
+                            is_argument = true;
+                        }
+                    }
+                }
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:1265]: This whole code here should be refactored into a function that should be put into ASRUtils. We might already have a function for this.

**Comment by certik** [comment]
Fix the above comment, and add a test.

**Comment by anutosh491** [comment]
I lack some experience working on shared branches but I was aware of how to use a branch (as in use the commits from it) and then build on top of it and finally raise a PR for the same. Hence I ended making #2453 which works as expected

**Comment by anutosh491** [comment]
I now realized why even after pushing on the same branch, it created a new PR.
So basically earlier as written in the documentation (https://docs.lfortran.org/en/installation/) we end up using 
`git clone https://github.com/lfortran/lfortran.git` and we don't end up cloning our fork, hence I had 
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ git remote -v
upstream  https://github.com/anutosh491/lpython.git (fetch)
upstream  https://github.com/anutosh491/lpython.git (push)
origin        https://github.com/lcompilers/lpython.git (fetch)
origin        https://github.com/lcompilers/lpython.git (push)
```

I realized long back but I made the switch just today to something like 
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ git remote -v
origin  https://github.com/anutosh491/lpython.git (fetch)
origin  https://github.com/anutosh491/lpython.git (push)
upstream        https://github.com/lcompilers/lpython.git (fetch)
upstream        https://github.com/lcompilers/lpython.git (push)
```
And though I can use my previous branches, github would end up creating a new branch out of then as I switched my origin. Luckily I don't have lot of open branches :)

**Comment by certik** [comment]
Just use `origin        https://github.com/lcompilers/lpython.git` and `anutosh https://github.com/anutosh491/lpython.git`. Always push to anutosh, never to origin.

**Pull request closed without merging**